### PR TITLE
Implement PLOOKup gate 

### PR DESCRIPTION
**Pull request by Bounce23**
This PR will focus on the standard composer, and will implement the lookup gate for construction in zk proofs.

### CODE REVIEW

**Comment by Bounce23** [comment]
This PR is ready for review. It the base for PR #334, thus perhaps it is better to have that merged in first.

**Review by ZER0 — CHANGES_REQUESTED** [review]
Some changes need to be done before proceed, see the comments.

File: plookup/src/table/lookup_table.rs [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+        let mut table: Vec<[BlsScalar; 3]> = Vec::with_capacity(cap);
+
+        for a in range.clone() {
+            range
+                .clone()
+                .map(|b| {
+                    let c = (a + b) % upper_bound;
+                    [BlsScalar::from(a), BlsScalar::from(b), BlsScalar::from(c)]
+                })
+                .for_each(|row| {
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/table/lookup_table.rs:57]: This function has to push each row and I tried to `collect` but for some reason the trait isn't implemented and I can't figure away to do it with the iterator.

File: plookup/src/multiset.rs [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    }
+
+    /// Concatenates and sorts two Multisets together.
+    /// From the Plookup paper, if we have t: {1,2,4,3}
+    /// and f: {2,3,4,1}.
+    /// We first check if all elements of f exist in t
+    /// Then we combine the multisets together and sort
+    /// their elements together. The final MultiSet will
+    /// look as follows, s: {1,1,2,2,3,3,4,4}
+    pub fn sort_and_index(&self, f: &MultiSet) -> Result<MultiSet, PlookupErrors> {
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/multiset.rs:67]: Name change done

File: plookup/src/multiset.rs [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+    /// Checks if an element is in the MultiSet
+    pub fn contains(&self, entry: &BlsScalar) -> bool {
+        self.0.contains(entry)
+    }
+
+    /// Splits a multiset into halves as specified by the paper
+    /// The last element of the first half should be the same
+    /// as the first element of the second half.
+    pub fn halve(&self) -> (MultiSet, MultiSet) {
+        let length = self.0.len();
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/multiset.rs:105]: It cannot, this is added in the doc comments

File: plookup/src/multiset.rs [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+            s.0.insert(index, *element);
+        }
+
+        Ok(s)
+    }
+
+    /// Checks whether one mutltiset is a subset of another.
+    /// This function will be used to check if the all elements
+    /// in set f, from the paper, are contained inside t.
+    pub fn is_subset_of(&self, other: &MultiSet) -> bool {
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/multiset.rs:83]: Amended

File: plookup/src/multiset.rs [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+
+    /// Concatenates and sorts two Multisets together.
+    /// From the Plookup paper, if we have t: {1,2,4,3}
+    /// and f: {2,3,4,1}.
+    /// We first check if all elements of f exist in t
+    /// Then we combine the multisets together and sort
+    /// their elements together. The final MultiSet will
+    /// look as follows, s: {1,1,2,2,3,3,4,4}
+    pub fn sort_and_index(&self, f: &MultiSet) -> Result<MultiSet, PlookupErrors> {
+        let _s: Vec<BlsScalar> = Vec::with_capacity(self.0.len() + f.0.len());
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/multiset.rs:68]: Amended

File: plookup/src/multiset.rs [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+        self.0.push(value)
+    }
+
+    /// Fetches last element in MultiSet.
+    /// Returns None if there are no elements in the MultiSet.
+    pub fn last(&self) -> Option<&BlsScalar> {
+        self.0.last()
+    }
+
+    fn from_slice(slice: &[BlsScalar]) -> MultiSet {
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/multiset.rs:45]: Done

File: plookup/src/table/witness_table.rs [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        self.f_1.push(left_wire_val);
+        self.f_2.push(right_wire_val);
+        self.f_3.push(output_wire_val);
+        Ok(())
+    }
+}
+
+impl WitnessTable4Arity {
+    /// Initialses empty witness table of arity 4
+    pub fn new() -> Self {
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/table/witness_table.rs:66]: Done

File: plookup/src/table/witness_table.rs [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+pub struct WitnessTable4Arity {
+    pub f_1: MultiSet,
+    pub f_2: MultiSet,
+    pub f_3: MultiSet,
+    pub f_4: MultiSet,
+}
+
+impl WitnessTable3Arity {
+    /// Initialises an empty witness table of arity 4
+    pub fn new() -> Self {
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/table/witness_table.rs:26]: Done

File: plookup/src/multiset.rs [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+    /// Creates an empty vector with a multiset wrapper around it
+    pub fn new() -> MultiSet {
+        MultiSet(vec![])
+    }
+
+    /// Extends the length of the multiset to 2^n elements
+    /// The 2^n will be the size of the arithmetic circuit.
+    /// This will extend the vectors to the size
+    pub fn pad(&mut self, n: u32) {
+        assert!(n.is_power_of_two());
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/multiset.rs:28]: It is not for debugging, so not to use an `assert!` in normal code, we could have a result returned, or leave it as it is?

File: plookup/src/table/lookup_table.rs [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+        Ok(self.0[pos][2])
+    }
+}
+
+/// This is a table, either
+pub struct PlookupTable4Arity(Vec<[BlsScalar; 4]>);
+
+impl PlookupTable4Arity {
+    /// Create a new, empty Plookup table, with arity 4.
+    pub fn new() -> Self {
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/table/lookup_table.rs:184]: Done

File: plookup/src/multiset.rs [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+    pub fn last(&self) -> Option<&BlsScalar> {
+        self.0.last()
+    }
+
+    fn from_slice(slice: &[BlsScalar]) -> MultiSet {
+        MultiSet(slice.to_vec())
+    }
+
+    /// Returns the cardinality of the multiset
+    pub fn len(&self) -> usize {
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/multiset.rs:50]: Done

File: plookup/src/multiset.rs [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+
+    /// Checks whether one mutltiset is a subset of another.
+    /// This function will be used to check if the all elements
+    /// in set f, from the paper, are contained inside t.
+    pub fn is_subset_of(&self, other: &MultiSet) -> bool {
+        let mut is_subset = true;
+
+        for x in self.0.iter() {
+            is_subset = other.contains(x);
+            if is_subset == false {
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/multiset.rs:88]: Done with a `contains all`

File: plookup/src/multiset.rs [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+
+/// MultiSet is struct containing vectors of scalars, which
+/// individually represents either a wire value or an index
+/// of a PlookUp table
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct MultiSet(Vec<BlsScalar>);
+
+impl MultiSet {
+    /// Creates an empty vector with a multiset wrapper around it
+    pub fn new() -> MultiSet {
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/src/multiset.rs:20]: Done

File: plookup/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+rand = "0.7.0"
+
+[package.metadata.docs.rs]
+rustdoc-args = [ "--html-in-header", "katex-header.html" ]
+
+[features]
+nightly = []
+trace = []
+trace-print = ["trace"]
+canon = ["dusk-bls12_381/canon", "dusk-jubjub/canon", "canonical"]
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/Cargo.toml:45]: Removed

File: plookup/Cargo.toml [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+rayon = "1.3.0"
+anyhow = "1.0.32"
+dusk-jubjub = "0.5.0"
+thiserror = "1.0"
+serde = "1.0"
+# Dusk related deps for WASMI serde
+canonical = {version = "0.4", optional = true}
+
+[dev-dependencies]
+rand = "0.7.0"
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/Cargo.toml:36]: Amended

File: plookup/Cargo.toml [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+dusk-bls12_381 = "0.3.0"
+itertools = "0.9.0"
+rand_chacha = "0.2"
+rayon = "1.3.0"
+anyhow = "1.0.32"
+dusk-jubjub = "0.5.0"
+thiserror = "1.0"
+serde = "1.0"
+# Dusk related deps for WASMI serde
+canonical = {version = "0.4", optional = true}
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/Cargo.toml:33]: Amended

File: plookup/Cargo.toml [hunk 17]
```diff
@@ -0,0 +1,45 @@
+[package]
+name = "dusk-plookup"
+version = "0.1.0"
+authors = ["Luke Pearson <luke@dusk.network>"]
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/Cargo.toml:4]: He has now been added

File: .gitignore [hunk 18]
```diff
@@ -6,4 +6,5 @@ scrap.rs
 kzg10.rs
 design.md
 refactor_notes
-.DS_Store
\ No newline at end of file
+.DS_Store
+.idea/
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply .gitignore:10]: Amended and this is now ignored only in the main `cargo.toml`

File: plookup/.gitignore [hunk 19]
```diff
@@ -0,0 +1,10 @@
+/target
+**/*.rs.bk
+Cargo.lock
+notes.md
+scrap.rs
+kzg10.rs
+design.md
+refactor_notes
+.DS_Store
+.idea/
```
[the remark this answers is not in the document]
Comment (Bounce23) [reply plookup/.gitignore:10]: As below

**Comment by CPerezz** [comment]
If this PR is going to be merged, commits like 2742337 shouldn't exist at all. Make sure to rebase and clean this branch as much as possible by removing all of the non-meaningful commits

**Comment by ZER0** [comment]
As discussed with @Bounce23, this PR is not longer valid and is going to be closed as wontfix.

**Pull request closed without merging**